torch/csrc/jit/codegen/cuda/scheduler/transpose.cpp

```diff
@@ -0,0 +1,637 @@
+  params.lparams.bind(warp_size, ParallelType::TIDx);
+  params.lparams.bind(kThreadsPerBlock / warp_size, ParallelType::TIDy);
+
+  if (isDebugDumpEnabled(DebugDumpOption::SchedulerDebug)) {
+    std::cerr << "\n===== Transpose Stats ========\n"
+              << "num_elems: " << n_elems << "\n"
+              << "n_input_tensors: " << n_input_tensors << "\n"
+              << "max_input_dtype_size: " << max_input_dtype_size << "\n"
+              << "group 1: " << ir_utils::toString(grouped_inputs_outputs[0])
+              << "\n"
+              << "group 2: " << ir_utils::toString(grouped_inputs_outputs[1])
+              << std::endl;
+    std::cerr << std::endl;
+    std::cerr << params.toString() << std::endl;
+  }
+
+  return params;
+}
+
+// TODO: remove or return launch parameters
+LaunchParams scheduleTranspose(
+    Fusion* fusion,
+    const at::ArrayRef<c10::IValue>& runtime_inputs) {
+  FUSER_PERF_SCOPE("scheduleFusion");
+  auto params = getTransposeHeuristics(fusion, runtime_inputs);
+  TORCH_INTERNAL_ASSERT(
+      params.has_value(), "Could not schedule pointwise operation.");
+  scheduleTranspose(fusion, params.value());
+  return params.value().lparams;
+}
+
+void scheduleTranspose(Fusion* fusion, const TransposeParams& params) {
+  FusionGuard fg(fusion);
+
+  // Make sure we don't have global memory set on intermediate tensors from
+  // fusion segmentation
+  scheduler_utils::clearMemorySpace(fusion);
+
+  // maybe has_reduction for scheduling should be done on a per output tensor
+  // basis.
+  TORCH_INTERNAL_ASSERT(
+      ir_utils::getReductionOps(fusion /*, ignore_trivial=true */).empty(),
+      "This scheduler only handles pointwise ops.");
+
+  // Cache inputs
+  auto cached_inputs = scheduler_utils::cacheInputs(fusion, true);
+
+  // Cache and fork outputs
+  auto cached_outputs = scheduler_utils::cacheAndForkOutputs(fusion, true);
+
+  std::vector<TensorView*> input_tvs;
+  {
+    auto filtered_tvs = ir_utils::filterByType<TensorView>(fusion->inputs());
+    // Remove hanging tensor views
+    for (auto tv : filtered_tvs) {
+      if (tv->uses().empty()) {
+        continue;
+      }
+      input_tvs.push_back(tv);
+    }
+  }
+  auto output_tvs = ir_utils::filterByType<TensorView>(fusion->outputs());
+
+  size_t max_dims = 0;
+  for (auto inp : input_tvs) {
+    max_dims = std::max(pointwise_utils::nRootDims(inp), max_dims);
+  }
+
+  for (auto out : output_tvs) {
+    max_dims = std::max(pointwise_utils::nRootDims(out), max_dims);
+  }
+
+  // If everything is zero dim tensors, just return.
+  if (max_dims == 0) {
+    return;
+  }
+
+  DomainMap domain_map(fusion);
+  auto grouped_inputs_outputs = domain_map.groupInputsOutputsByInnerDim();
+  TORCH_INTERNAL_ASSERT(grouped_inputs_outputs.size() >= 2);
+
+  // We need something similar to `cacheFork` for input tensors in group 2. We
+  // need this because we will want to propagate to the entire DAG except group
```

Comment:
Which propagation needs this?

Reply:
Are caching inputs and outputs of group two so you can make shared memory tiles for those caches, do the transposes there,  then everything else in the graph should be scheduled as one big group one?

Reply:
What you said in https://github.com/csarofeen/pytorch/pull/1854#r928162687 is almost right, except for the "two steps" vs "three steps" thing https://github.com/csarofeen/pytorch/pull/1854#r929503588. So yes, the caches are stored in shared memory, and I do the transpose there, and everything else should be scheduled as group 1. And the propagation that needs this is the last step, i.e. "reschedule based on reference 1 then propagate that information".